Add separate warnings for settings that depend on illink pack

sbomer opened the pull request:
This splits the error added in https://github.com/dotnet/sdk/pull/33041 into separate warnings/errors specific to the various options that depend on the ILLink pack. These warnings/errors are produced when those options are set for an unsupported target framework.

The PublishTrimmed error text reuses the message that used to be produced as a warning when targeting < netcoreapp3.0. This error is produced only when PublishAot is not set (there's a more specific error if it is).

Previously, we would produce errors when attempting to use the analyzers on unsupported TFMs. This has been changed to a warning in the case of a library (for example IsTrimmable, rather than PublishTrimmed). See comments in ProcessFrameworkReferences for details.

This also fixes an issue where ProcessFrameworkReferences was not running for netcoreapp2.0 because there were no FrameworkReferences defined for that TFM. This adds an extra condition to let ProcessFrameworkReferences run if any settings depend on the ILLink pack, so that it can produce proper warnings/errors.

Fixes https://github.com/dotnet/linker/issues/3175

sbomer committed:
ProcessFrameworkReferences if tool packs are required

sbomer committed:
Fix test failures

- Don't publish single-file test for unsupported net472 TFM
- Expect single-file to fail adding ILLink pack reference for unupported TFMs
- Adjust TFM to check single-file error behavior when publishing library

sbomer committed:
Fix another test

sbomer committed:
Fix another test

sbomer committed:
Merge remote-tracking branch 'origin/main' into processFrameworkReferencesCondition

sbomer committed:
Fix build

sbomer committed:
Limit scope of change to illink pack

sbomer committed:
Add separate warnings for settings that depend on illink pack

sbomer committed:
Fix some tests

sbomer committed:
Add IsTrimmable test, fix PublishTrimmed test

sbomer committed:
Clean up 'Error' from warning id

sbomer committed:
Cleanup

sbomer committed:
Name string ids consistently

sbomer committed:
Merge branch 'processFrameworkReferencesCondition' into improveTrimmingWarnings

sbomer committed:
Fix build

sbomer committed:
Remove existing error path

sbomer committed:
Consolidate tests, clean up condition

sbomer committed:
More cleanup

sbomer committed:
Add tests for new warnings

Also fix some issues with existing tests

sbomer committed:
Don't consider PublishSingleFile

PublishSingleFile without EnableSingleFileAnalyzer
doesn't require the ILLink pack.

Also add some comments, clean up
some logic.

sbomer committed:
Merge remote-tracking branch 'origin/main' into improveTrimmingWarnings

sbomer committed:
Revert single-file test changes

vitek-karas commented on src/Tasks/Common/Resources/Strings.resx:
```diff
@@ -937,5 +939,14 @@ You may need to build the project on another operating system or architecture, o
-  <!-- The latest message added is UnsupportedVisualStudioVersion. Please update this value with each PR to catch parallel PRs both adding a new message -->
+  <data name="IsAotCompatibleUnsupported" xml:space="preserve">
+    <value>NETSDK1210: IsAotCompatible is not supported for the target framework. Consider multi-targeting to a supported framework to enable ahead-of-time compilation analysis, and set IsAotCompatible only for the supported frameworks. For example:
+&lt;IsAotCompatible Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net6.0'))"&gt;true&lt;/IsAotCompatible&gt;</value>
```
AOT is really only supported from 7 and above - so the example should use `net7.0`.

vitek-karas commented on src/Tasks/Common/Resources/Strings.resx:
```diff
@@ -937,5 +939,14 @@ You may need to build the project on another operating system or architecture, o
+    <comment>{StrBegin="NETSDK1210: "}</comment>
+  </data>
+  <data name="EnableSingleFileAnalyzerUnsupported" xml:space="preserve">
+    <value>NETSDK1211: EnableSingleFileAnalyzer is not supported for the target framework.</value>
```
Maybe also add the example with the conditional syntax - since we don't have an `IsSingleFileCompatible` the `EnableSingleFileAnalyzer` effectively acts like it - so the guidance should now apply to it.

vitek-karas reviewed (approved):
Looks good to me - just minor comments

sbomer committed:
PR feedback

- net7.0 for AOT
- Add IsTargetFrameworkCompatible to message for
  single-file analysis

sbomer committed:
Fix test

sbomer committed:
Merge remote-tracking branch 'origin/main' into improveTrimmingWarnings
